sycl/include/CL/sycl/ONEAPI/reduction.hpp

```diff
@@ -665,105 +729,105 @@ class reduction_impl : private reduction_impl_base {
+      return *MRWAcc;
 
-    // Create a new output buffer and return an accessor to it.
-    MOutBufPtr = std::make_shared<buffer<T, buffer_dim>>(range<1>(Size));
-    accessor_type NewAcc(*MOutBufPtr);
+    auto RWReduVal = std::make_shared<T>(MIdentity);
+    CGH.addReduction(RWReduVal);
+    MOutBufPtr = std::make_shared<buffer<T, 1>>(RWReduVal.get(), range<1>(1));
     CGH.addReduction(MOutBufPtr);
-    CGH.require(NewAcc);
-    return NewAcc;
+    return createHandlerWiredReadWriteAccessor(CGH, *MOutBufPtr);
   }
 
-  /// Creates 1-element global buffer initialized with identity value and
-  /// returns an accessor to that buffer.
-  rw_accessor_type getReadWriteScalarAcc(handler &CGH) const {
-    auto RWReduVal = std::make_shared<T>(MIdentity);
-    CGH.addReduction(RWReduVal);
-    auto RWReduBuf =
-        std::make_shared<buffer<T, 1>>(RWReduVal.get(), range<1>(1));
-    CGH.addReduction(RWReduBuf);
-    return accessor<T, Dims, access::mode::read_write,
-                    access::target::global_buffer>(*RWReduBuf, CGH);
+  bool hasUserDiscardWriteAccessor() { return MDWAcc != nullptr; }
+
+  template <bool _IsUSM = IsUSM>
+  std::enable_if_t<!_IsUSM, rw_accessor_type &> getUserReadWriteAccessor() {
+    return *MRWAcc;
   }
 
-  accessor_type &getUserAccessor() { return *MAcc; }
+  template <bool _IsUSM = IsUSM>
+  std::enable_if_t<!_IsUSM, dw_accessor_type &> getUserDiscardWriteAccessor() {
+    return *MDWAcc;
+  }
 
-  T *getUSMPointer() {
+  result_type *getUSMPointer() {
     assert(is_usm && "Unexpected call of getUSMPointer().");
     return MUSMPointer;
   }
 
-  template <typename AccT>
-  enable_if_t<std::is_same<AccT, rw_accessor_type>::value ||
-                  std::is_same<AccT, accessor_type>::value,
-              result_type *> static inline getOutPointer(const AccT &OutAcc) {
+  static inline result_type *getOutPointer(const rw_accessor_type &OutAcc) {
     return OutAcc.get_pointer().get();
   }
 
-  template <typename ResT>
-  enable_if_t<std::is_same<ResT, result_type>::value,
-              result_type *> static inline getOutPointer(ResT *OutPtr) {
+  static inline result_type *getOutPointer(result_type *OutPtr) {
     return OutPtr;
   }
 
   /// Returns the binary operation associated with the reduction.
   BinaryOperation getBinaryOperation() const { return MBinaryOp; }
+  bool initializeToIdentity() const { return InitializeToIdentity; }
 
 private:
+  template <typename BufferT, access::placeholder IsPH = IsPlaceholder>
+  std::enable_if_t<IsPH == access::placeholder::false_t, rw_accessor_type>
+  createHandlerWiredReadWriteAccessor(handler &CGH, BufferT Buffer) {
+    return {Buffer, CGH};
+  }
+
+  template <typename BufferT, access::placeholder IsPH = IsPlaceholder>
+  std::enable_if_t<IsPH == access::placeholder::true_t, rw_accessor_type>
+  createHandlerWiredReadWriteAccessor(handler &CGH, BufferT Buffer) {
+    rw_accessor_type Acc(Buffer);
+    CGH.require(Acc);
+    return Acc;
+  }
+
```

Comment:
It wasn't immediately obvious to me why the body of these two functions is so different.  Is this just because calling `require` on a non-placeholder accessor is redundant?

Reply:
The require() method is available only for placeholder accessor.
non-placeholder accessor must be constructed with CGH parameter. Placeholder accessors cannot accept CGH to ctor.

Reply:
Makes sense.  Thanks!